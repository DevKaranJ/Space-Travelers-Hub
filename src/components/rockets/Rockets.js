import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import {
  fetchRockets,
  reserveRocket,
} from '../../redux/features/rockets/rocketsSlice';

const Rockets = () => {
  const rocketsState = useSelector((state) => state.rockets.rockets);
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(fetchRockets());
  }, [dispatch]);

  const handleReserveClick = (rocket) => {
    dispatch(reserveRocket(rocket.id));
    const cancelReservationBtn = document.querySelector('.reserve-btn');
    cancelReservationBtn.innerHTML = 'Cancel Reservation';
  };

  return (
    <section className="rockets-container">
      <ul className="rocket-info-container">
        {rocketsState.map((rocket) => (
          <li key={rocket.id}>
            {rocket.flickr_images.length > 0 && (
              <img
                src={rocket.flickr_images[0]}
                alt={`Rocket ${rocket.name} Imge`}
                className="rocket-image"
              />
            )}
            <div className="rocket-info">
              <h2>{rocket.name}</h2>
              <p className="reserve-tag">
                {rocket.reserved && (
                  <span className="reserve-rocket-tag">Reserved</span>
                )}
                {rocket.description}
              </p>
              <button
                type="button"
                className="reserve-btn"
                onClick={() => handleReserveClick(rocket)}
              >
                Reserve rocket
              </button>
            </div>
          </li>
        ))}
      </ul>
    </section>
  );
};

export default Rockets;
