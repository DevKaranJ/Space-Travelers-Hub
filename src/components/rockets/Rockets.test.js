import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import Rockets from './Rockets';

const middlewares = [thunk];

describe('Rockets Component', () => {
  const initialState = {
    rockets: {
      rockets: [
        {
          id: '1',
          name: 'Falcon 1',
          flickr_images: ['image-url-1'],
          description: 'Description 1',
          reserved: false,
        },
        {
          id: '2',
          name: 'Falcon 9',
          flickr_images: ['image-url-2'],
          description: 'Description 2',
          reserved: true,
        },
      ],
    },
  };

  const mockStore = configureMockStore(middlewares);
  let store;

  beforeEach(() => {
    store = mockStore(initialState);
  });

  test('Rockets component renders correctly', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <Rockets />
      </Provider>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test('Rockets component displays rockets correctly', () => {
    render(
      <Provider store={store}>
        <Rockets />
      </Provider>,
    );

    // Test that Rocket 1 is displayed
    expect(screen.getByText('Falcon 1')).toBeInTheDocument();
    expect(screen.getByText('Reserve rocket')).toBeInTheDocument();
    expect(screen.getByText('Cancel Reservation')).toBeInTheDocument();

    // Test that Rocket 2 is displayed
    expect(screen.getByText('Falcon 9')).toBeInTheDocument();
    expect(screen.getByText('Reserve rocket')).toBeInTheDocument();
    expect(screen.getByText('Cancel Reservation')).toBeInTheDocument();
  });
});
