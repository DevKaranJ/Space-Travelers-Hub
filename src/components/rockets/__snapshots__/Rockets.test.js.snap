// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rockets Component Rockets component renders correctly 1`] = `
<DocumentFragment>
  <section
    class="rockets-container"
  >
    <ul
      class="rocket-info-container"
    >
      <li>
        <img
          alt="Rocket Falcon 1 Imge"
          class="rocket-image"
          src="image-url-1"
        />
        <div
          class="rocket-info"
        >
          <h2>
            Falcon 1
          </h2>
          <p
            class="reserve-tag"
          >
            Description 1
          </p>
          <button
            class="reserve-btn"
            type="button"
          >
            Reserve rocket
          </button>
        </div>
      </li>
      <li>
        <img
          alt="Rocket Falcon 9 Imge"
          class="rocket-image"
          src="image-url-2"
        />
        <div
          class="rocket-info"
        >
          <h2>
            Falcon 9
          </h2>
          <p
            class="reserve-tag"
          >
            <span
              class="reserve-rocket-tag"
            >
              Reserved
            </span>
            Description 2
          </p>
          <button
            class="reserve-btn cancel-reserve-btn"
            type="button"
          >
            Cancel Reservation
          </button>
        </div>
      </li>
    </ul>
  </section>
</DocumentFragment>
`;
